File: skills/realtime/websocket-implementation/resources/examples/python/websocket_server.py
#!/usr/bin/env python3
"""
WebSocket Server Example (Python)

Demonstrates a production-ready WebSocket server using the 'websockets' library.
Features: broadcast, rooms, authentication, async/await patterns.

Installation:
    pip install websockets

Usage:
    python websocket_server.py
    python websocket_server.py --port 8080
"""

import asyncio
import json
import logging
import signal
import time
from collections import defaultdict
from dataclasses import dataclass, asdict
from typing import Set, Dict, Optional
import websockets
from websockets.server import WebSocketServerProtocol


# Configuration
PORT = 8080
PING_INTERVAL = 30  # seconds
PING_TIMEOUT = 5    # seconds

# Logging
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s [%(levelname)s] %(message)s'
)
logger = logging.getLogger(__name__)


@dataclass
class Metrics:
    """Server metrics"""
    connections: int = 0
    total_connections: int = 0
    messages_received: int = 0
    messages_sent: int = 0
    bytes_received: int = 0
    bytes_sent: int = 0


class RateLimiter:
    """Simple rate limiter using token bucket algorithm"""

    def __init__(self, max_rate: int = 10):
        self.max_rate = max_rate
        self.buckets: Dict[WebSocketServerProtocol, dict] = {}

    def check(self, ws: WebSocketServerProtocol) -> bool:
        """Check if request is within rate limit"""
        if ws not in self.buckets:
            self.buckets[ws] = {
                'count': 0,
                'reset_time': time.time() + 1.0
            }

        bucket = self.buckets[ws]

        if time.time() > bucket['reset_time']:
            bucket['count'] = 0
            bucket['reset_time'] = time.time() + 1.0

        bucket['count'] += 1

        return bucket['count'] <= self.max_rate

    def cleanup(self, ws: WebSocketServerProtocol):
        """Remove bucket for disconnected client"""
        self.buckets.pop(ws, None)


class WebSocketServer:
    """WebSocket server with rooms and broadcast support"""

    def __init__(self):
        self.clients: Set[WebSocketServerProtocol] = set()
        self.rooms: Dict[str, Set[WebSocketServerProtocol]] = defaultdict(set)
        self.metrics = Metrics()
        self.rate_limiter = RateLimiter(max_rate=10)

    async def register(self, ws: WebSocketServerProtocol):
        """Register a new client"""
        self.clients.add(ws)
        self.metrics.connections += 1
        self.metrics.total_connections += 1

        ws.user_id = f"user_{id(ws) % 10000:04d}"
        ws.room = None

        logger.info(f"Client connected: {ws.user_id} ({ws.remote_address})")

        # Send welcome message
        await self.send_message(ws, {
            'type': 'welcome',
            'user_id': ws.user_id,
            'timestamp': time.time()
        })

    async def unregister(self, ws: WebSocketServerProtocol):
        """Unregister a client"""
        self.clients.discard(ws)
        self.metrics.connections -= 1

        # Leave room if in one
        if hasattr(ws, 'room') and ws.room:
            await self.leave_room(ws)

        # Cleanup rate limiter
        self.rate_limiter.cleanup(ws)

        logger.info(f"Client disconnected: {ws.user_id}")

    async def join_room(self, ws: WebSocketServerProtocol, room: str):
        """Join a room"""
        # Leave current room if in one
        if hasattr(ws, 'room') and ws.room:
            await self.leave_room(ws)

        # Join new room
        self.rooms[room].add(ws)
        ws.room = room

        logger.info(f"Client {ws.user_id} joined room: {room} ({len(self.rooms[room])} members)")

        # Notify room members
        await self.broadcast_to_room(room, {
            'type': 'user_joined',
            'user_id': ws.user_id,
            'room': room,
            'timestamp': time.time()
        }, exclude=ws)

        # Confirm to client
        await self.send_message(ws, {
            'type': 'joined',
            'room': room,
            'members': len(self.rooms[room]),
            'timestamp': time.time()
        })

    async def leave_room(self, ws: WebSocketServerProtocol):
        """Leave current room"""
        if not hasattr(ws, 'room') or not ws.room:
            return

        room = ws.room
        self.rooms[room].discard(ws)

        # Notify room members
        await self.broadcast_to_room(room, {
            'type': 'user_left',
            'user_id': ws.user_id,
            'room': room,
            'timestamp': time.time()
        })

        # Clean up empty room
        if len(self.rooms[room]) == 0:
            del self.rooms[room]

        logger.info(f"Client {ws.user_id} left room: {room}")

        ws.room = None

        # Confirm to client
        await self.send_message(ws, {
            'type': 'left',
            'timestamp': time.time()
        })

    async def send_message(self, ws: WebSocketServerProtocol, message: dict):
        """Send message to a single client"""
        try:
            data = json.dumps(message)
            await ws.send(data)
            self.metrics.messages_sent += 1
            self.metrics.bytes_sent += len(data)
        except Exception as e:
            logger.error(f"Error sending message to {ws.user_id}: {e}")

    async def broadcast_to_room(self, room: str, message: dict, exclude: Optional[WebSocketServerProtocol] = None):
        """Broadcast message to all clients in a room"""
        if room not in self.rooms:
            return

        data = json.dumps(message)
        tasks = []

        for client in self.rooms[room]:
            if client != exclude:
                tasks.append(client.send(data))
                self.metrics.messages_sent += 1
                self.metrics.bytes_sent += len(data)

        if tasks:
            await asyncio.gather(*tasks, return_exceptions=True)

    async def broadcast_all(self, message: dict, exclude: Optional[WebSocketServerProtocol] = None):
        """Broadcast message to all connected clients"""
        data = json.dumps(message)
        tasks = []

        for client in self.clients:
            if client != exclude:
                tasks.append(client.send(data))
                self.metrics.messages_sent += 1
                self.metrics.bytes_sent += len(data)

        if tasks:
            await asyncio.gather(*tasks, return_exceptions=True)

    async def handle_message(self, ws: WebSocketServerProtocol, data: str):
        """Handle incoming message"""
        self.metrics.messages_received += 1
        self.metrics.bytes_received += len(data)

        # Rate limiting
        if not self.rate_limiter.check(ws):
            await self.send_message(ws, {
                'type': 'error',
                'error': 'Rate limit exceeded'
            })
            return

        # Parse message
        try:
            msg = json.loads(data)
        except json.JSONDecodeError:
            await self.send_message(ws, {
                'type': 'error',
                'error': 'Invalid JSON'
            })
            return

        msg_type = msg.get('type')

        # Route message
        if msg_type == 'join':
            room = msg.get('room')
            if room:
                await self.join_room(ws, room)
            else:
                await self.send_message(ws, {
                    'type': 'error',
                    'error': 'Missing room parameter'
                })

        elif msg_type == 'leave':
            await self.leave_room(ws)

        elif msg_type == 'message':
            if not hasattr(ws, 'room') or not ws.room:
                await self.send_message(ws, {
                    'type': 'error',
                    'error': 'Not in a room'
                })
                return

            content = msg.get('content')
            if content:
                await self.broadcast_to_room(ws.room, {
                    'type': 'message',
                    'user_id': ws.user_id,
                    'room': ws.room,
                    'content': content,
                    'timestamp': time.time()
                }, exclude=ws)

                logger.info(f"Message in room {ws.room} from {ws.user_id}: {content}")

        elif msg_type == 'broadcast':
            content = msg.get('content')
            if content:
                await self.broadcast_all({
                    'type': 'broadcast',
                    'user_id': ws.user_id,
                    'content': content,
                    'timestamp': time.time()
                }, exclude=ws)

                logger.info(f"Broadcast from {ws.user_id}: {content}")

        elif msg_type == 'ping':
            await self.send_message(ws, {
                'type': 'pong',
                'timestamp': time.time()
            })

        else:
            await self.send_message(ws, {
                'type': 'error',
                'error': f'Unknown message type: {msg_type}'
            })

    async def handler(self, ws: WebSocketServerProtocol, path: str):
        """Main WebSocket handler"""
        await self.register(ws)

        try:
            async for message in ws:
                await self.handle_message(ws, message)

        except websockets.exceptions.ConnectionClosed:
            pass
        except Exception as e:
            logger.error(f"Error in handler for {ws.user_id}: {e}")
        finally:
            await self.unregister(ws)

    async def start(self, host: str = 'localhost', port: int = PORT):
        """Start the WebSocket server"""
        async with websockets.serve(
            self.handler,
            host,
            port,
            ping_interval=PING_INTERVAL,
            ping_timeout=PING_TIMEOUT
        ):
            logger.info(f"WebSocket server running on ws://{host}:{port}")

            # Metrics reporting
            async def report_metrics():
                while True:
                    await asyncio.sleep(60)
                    logger.info(
                        f"Metrics: connections={self.metrics.connections}, "
                        f"total={self.metrics.total_connections}, "
                        f"msgs_recv={self.metrics.messages_received}, "
                        f"msgs_sent={self.metrics.messages_sent}"
                    )

            asyncio.create_task(report_metrics())

            # Wait forever
            await asyncio.Future()


async def main():
    """Main entry point"""
    import argparse

    parser = argparse.ArgumentParser(description='WebSocket Server')
    parser.add_argument('--host', default='localhost', help='Host to bind to')
    parser.add_argument('--port', type=int, default=PORT, help='Port to bind to')
    args = parser.parse_args()

    server = WebSocketServer()

    # Graceful shutdown
    loop = asyncio.get_event_loop()

    def shutdown():
        logger.info("Shutting down...")
        loop.stop()

    for sig in (signal.SIGTERM, signal.SIGINT):
        loop.add_signal_handler(sig, shutdown)

    try:
        await server.start(host=args.host, port=args.port)
    except KeyboardInterrupt:
        logger.info("Interrupted by user")


if __name__ == '__main__':
    asyncio.run(main())

File: skills/realtime/websocket-implementation/resources/examples/node/websocket-server.js
/**
 * WebSocket Server Example (Node.js)
 *
 * Demonstrates a production-ready WebSocket server using the 'ws' library.
 * Features: broadcast, rooms, authentication, rate limiting, metrics.
 *
 * Installation:
 *   npm install ws
 *
 * Usage:
 *   node websocket-server.js
 *   node websocket-server.js --port 8080 --redis redis://localhost:6379
 */

const WebSocket = require('ws');
const http = require('http');
const crypto = require('crypto');

// Configuration
const PORT = process.env.PORT || 8080;
const PING_INTERVAL = 30000; // 30 seconds
const MESSAGE_RATE_LIMIT = 10; // messages per second
const MAX_MESSAGE_SIZE = 100 * 1024; // 100KB

// Metrics
const metrics = {
  connections: 0,
  totalConnections: 0,
  messagesReceived: 0,
  messagesSent: 0,
  bytesReceived: 0,
  bytesSent: 0
};

// Room management
const rooms = new Map();

// Rate limiting
class RateLimiter {
  constructor(maxRate) {
    this.maxRate = maxRate;
    this.buckets = new WeakMap();
  }

  check(ws) {
    if (!this.buckets.has(ws)) {
      this.buckets.set(ws, {
        count: 0,
        resetTime: Date.now() + 1000
      });
    }

    const bucket = this.buckets.get(ws);

    if (Date.now() > bucket.resetTime) {
      bucket.count = 0;
      bucket.resetTime = Date.now() + 1000;
    }

    bucket.count++;

    return bucket.count <= this.maxRate;
  }
}

const rateLimiter = new RateLimiter(MESSAGE_RATE_LIMIT);

// Broadcast to all clients in a room
function broadcastToRoom(room, message, sender = null) {
  const roomClients = rooms.get(room);
  if (!roomClients) return;

  const data = JSON.stringify(message);
  let sent = 0;

  roomClients.forEach((client) => {
    if (client !== sender && client.readyState === WebSocket.OPEN) {
      client.send(data);
      metrics.messagesSent++;
      metrics.bytesSent += data.length;
      sent++;
    }
  });

  return sent;
}

// Broadcast to all clients
function broadcastAll(message, sender = null) {
  const data = JSON.stringify(message);
  let sent = 0;

  wss.clients.forEach((client) => {
    if (client !== sender && client.readyState === WebSocket.OPEN) {
      client.send(data);
      metrics.messagesSent++;
      metrics.bytesSent += data.length;
      sent++;
    }
  });

  return sent;
}

// Join room
function joinRoom(ws, room) {
  if (!rooms.has(room)) {
    rooms.set(room, new Set());
  }

  rooms.get(room).add(ws);
  ws.room = room;

  console.log(`Client ${ws.id} joined room: ${room} (${rooms.get(room).size} members)`);

  // Notify room
  broadcastToRoom(room, {
    type: 'user_joined',
    userId: ws.userId,
    room: room,
    timestamp: Date.now()
  }, ws);
}

// Leave room
function leaveRoom(ws) {
  if (ws.room && rooms.has(ws.room)) {
    rooms.get(ws.room).delete(ws);

    // Notify room
    broadcastToRoom(ws.room, {
      type: 'user_left',
      userId: ws.userId,
      room: ws.room,
      timestamp: Date.now()
    }, ws);

    // Clean up empty rooms
    if (rooms.get(ws.room).size === 0) {
      rooms.delete(ws.room);
    }

    console.log(`Client ${ws.id} left room: ${ws.room}`);
    ws.room = null;
  }
}

// Message handlers
const messageHandlers = {
  join: (ws, data) => {
    if (ws.room) {
      leaveRoom(ws);
    }
    joinRoom(ws, data.room);

    ws.send(JSON.stringify({
      type: 'joined',
      room: data.room,
      timestamp: Date.now()
    }));
  },

  leave: (ws) => {
    leaveRoom(ws);

    ws.send(JSON.stringify({
      type: 'left',
      timestamp: Date.now()
    }));
  },

  message: (ws, data) => {
    if (!ws.room) {
      ws.send(JSON.stringify({
        type: 'error',
        error: 'Not in a room'
      }));
      return;
    }

    const sent = broadcastToRoom(ws.room, {
      type: 'message',
      userId: ws.userId,
      room: ws.room,
      content: data.content,
      timestamp: Date.now()
    }, ws);

    console.log(`Message in room ${ws.room}: ${data.content} (sent to ${sent} clients)`);
  },

  ping: (ws) => {
    ws.send(JSON.stringify({
      type: 'pong',
      timestamp: Date.now()
    }));
  },

  broadcast: (ws, data) => {
    const sent = broadcastAll({
      type: 'broadcast',
      userId: ws.userId,
      content: data.content,
      timestamp: Date.now()
    }, ws);

    console.log(`Broadcast: ${data.content} (sent to ${sent} clients)`);
  }
};

// Create HTTP server (for health checks, metrics)
const server = http.createServer((req, res) => {
  if (req.url === '/health') {
    res.writeHead(200, { 'Content-Type': 'application/json' });
    res.end(JSON.stringify({ status: 'ok', uptime: process.uptime() }));
  } else if (req.url === '/metrics') {
    res.writeHead(200, { 'Content-Type': 'application/json' });
    res.end(JSON.stringify(metrics));
  } else {
    res.writeHead(404);
    res.end('Not Found');
  }
});

// Create WebSocket server
const wss = new WebSocket.Server({
  server,
  maxPayload: MAX_MESSAGE_SIZE,
  perMessageDeflate: {
    zlibDeflateOptions: {
      level: 6
    },
    threshold: 1024
  }
});

// Connection handler
wss.on('connection', (ws, req) => {
  // Generate connection ID
  ws.id = crypto.randomBytes(8).toString('hex');
  ws.userId = `user_${ws.id.slice(0, 8)}`;
  ws.isAlive = true;
  ws.room = null;

  metrics.connections++;
  metrics.totalConnections++;

  console.log(`Client connected: ${ws.id} (${req.socket.remoteAddress})`);

  // Send welcome message
  ws.send(JSON.stringify({
    type: 'welcome',
    userId: ws.userId,
    timestamp: Date.now()
  }));

  // Message handler
  ws.on('message', (data, isBinary) => {
    metrics.messagesReceived++;
    metrics.bytesReceived += data.length;

    // Rate limiting
    if (!rateLimiter.check(ws)) {
      ws.send(JSON.stringify({
        type: 'error',
        error: 'Rate limit exceeded'
      }));
      return;
    }

    // Parse message
    let msg;
    try {
      msg = isBinary ? data : JSON.parse(data.toString());
    } catch (e) {
      ws.send(JSON.stringify({
        type: 'error',
        error: 'Invalid JSON'
      }));
      return;
    }

    // Handle message
    const handler = messageHandlers[msg.type];
    if (handler) {
      try {
        handler(ws, msg);
      } catch (e) {
        console.error(`Error handling ${msg.type}:`, e);
        ws.send(JSON.stringify({
          type: 'error',
          error: 'Internal server error'
        }));
      }
    } else {
      ws.send(JSON.stringify({
        type: 'error',
        error: `Unknown message type: ${msg.type}`
      }));
    }
  });

  // Pong handler (for heartbeat)
  ws.on('pong', () => {
    ws.isAlive = true;
  });

  // Close handler
  ws.on('close', (code, reason) => {
    metrics.connections--;
    leaveRoom(ws);
    console.log(`Client disconnected: ${ws.id} (code: ${code}, reason: ${reason.toString()})`);
  });

  // Error handler
  ws.on('error', (error) => {
    console.error(`Client error (${ws.id}):`, error.message);
  });
});

// Heartbeat interval (ping/pong)
const heartbeatInterval = setInterval(() => {
  wss.clients.forEach((ws) => {
    if (ws.isAlive === false) {
      console.log(`Terminating inactive connection: ${ws.id}`);
      return ws.terminate();
    }

    ws.isAlive = false;
    ws.ping();
  });
}, PING_INTERVAL);

// Cleanup on server close
wss.on('close', () => {
  clearInterval(heartbeatInterval);
});

// Graceful shutdown
process.on('SIGTERM', () => {
  console.log('SIGTERM received, closing server...');

  wss.clients.forEach((ws) => {
    ws.close(1001, 'Server shutting down');
  });

  server.close(() => {
    console.log('Server closed');
    process.exit(0);
  });

  // Force exit after 10 seconds
  setTimeout(() => {
    console.error('Forced exit after timeout');
    process.exit(1);
  }, 10000);
});

// Start server
server.listen(PORT, () => {
  console.log(`WebSocket server running on ws://localhost:${PORT}`);
  console.log(`Health check: http://localhost:${PORT}/health`);
  console.log(`Metrics: http://localhost:${PORT}/metrics`);
});
